client/finality-grandpa/src/environment.rs

```diff
@@ -1077,6 +1081,24 @@ pub(crate) fn finalize_block<BE, Block, Client>(
 			},
 		};
 
+		if let Some(notifiers) = finality_notifiers {
+			// Q: We `finalized()` this at L37, so can I be sure
+			// that it's fine to unwrap here?
+			if let Some(justification) = justification.clone() {
+				let header = client.header(BlockId::Hash(hash))?
+					.expect("");
+				let notification = JustificationNotification {
+					header,
+					justification,
+				};
+
+				for s in notifiers.read().iter() {
```

Comment:
You probably want a `Mutex` instead and call `retain` to remove receivers that are no longer active.

Reply:
Best this should use the sending half of subscription manager (see other comments) and this whole retain/Mutex logic should stay encapsulated.